Add EIP: ERC-721 Multi-Metadata Extension (#7160)
ethereum/EIPs: EIPS/eip-7160.md

---
eip: 7160
title: ERC-721 Multi-Metadata Extension

> description: "This EIP proposes an extension to the ERC-721 to support multiple metadata URIs per token."

Reviewer: Can condense this a bit, and make it more descriptive:

```suggestion
description: Multiple metadata URIs per token, with a single pinned primary URI
```

author: 0xG (@0xGh), Marco Peyfuss (@mpeyfuss)
discussions-to: https://ethereum-magicians.org/t/erc721-multi-metadata-extension/14629
status: Draft
type: Standards Track
category: ERC
created: 2023-06-09
requires: 165, 721
---

## Abstract

This EIP proposes an extension to the [ERC-721](./eip-721.md) standard to support multiple metadata URIs per token. It introduces a new interface, `IERC721MultiMetadata`, which provides methods for accessing the metadata URIs associated with a token, including a pinned URI index and a list of all metadata URIs. The extension is designed to be backward compatible with existing `ERC721Metadata` implementations.

## Motivation

> The current [ERC-721](./eip-721.md) standard allows for a single metadata URI per token. However, there are use cases where multiple metadata URIs are desirable, such as when a token represents a collection of (cycling) assets with individual metadata, historic token metadata, collaborative and multi-artist tokens, evolving tokens. This extension enables such use cases by introducing the concept of multi-metadata support.

Reviewer: I think it would be important to include answers to:

 - Why does this need to be more than just `tokenURI` and some token-specific functions to change the active URI?
 - Who would be interested in the non-pinned URIs?
 - How would this information be displayed?

Author: Added a paragraph right below. Let me know what you think. Also @mpeyfuss can chime in and add further reasons for why having a multi-metadata standard might be a good idea.

Reviewer 2: Yeah there are quite a few examples we've found working with creators and collectors on why multi-metadata is interesting. Can add example use cases. One of the most mentioned reasons is that of having the option to provide versions of the token that match different aspect ratios and could be utilized for specific display purposes while keeping provenance/ownership intact.

## Specification

The `IERC721MultiMetadata` interface extends the existing `IERC721` interface and introduces additional methods and events:

```solidity

> interface IERC721MultiMetadata is IERC721 {

Author: I ended up not extending IERC721Metadata because it is not a standard with its own EIP and technically not needed - a token with tokenURIs alone would work if supported by marketplaces. 

The ERC721 standard suggest that IERC721Metadata must be implemented, and in fact the OpenZeppelin contract does that.

If we were to extend IERC721Metadata we would need to override tokenURI by specifying that it `override(ERC721, IERC721Metadata)`

Open to change my mind about this.

> interface IERC721MultiMetadata is IERC721 {

Reviewer: I'm no Solidity expert, but I've been told to avoid making interfaces depend on one another:

```suggestion
interface IERC721MultiMetadata /* is IERC721 */ {
```

Author: Commented out as suggested. By the way should the interface name reflect the EIP name instead of being called `IERC721MultiMetadata` or is it fine to give it a name?

Reviewer: You can name it however you like!

Depending on the implementation, the `pinTokenURI` function allows the contract owner or token owner to specify a particular fixed metadata URI index for a token. This enables the selection of a preferred URI by index from the list of available metadata.

When unpinned, it is recommended to return the last URI for the token. However the behavior in the case of unpinned tokens is at the discretion of the implementation and depends on the specific purpose of the token.

## Backwards Compatibility

This extension is designed to be backward compatible with existing [ERC-721](./eip-721.md) contracts. The implementation of the `tokenURI` method must either return the last URI in the list returned by tokenURIs or the pinned URI.

## Reference Implementation

An open-source reference implementation of the `IERC721MultiMetadata` interface can be provided, demonstrating how to extend an existing [ERC-721](./eip-721.md) contract to support multi-metadata functionality. This reference implementation can serve as a guide for developers looking to implement the extension in their own contracts.

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: Code included inline in the EIP must be `CC0-1.0` licensed. Code in the `assets/...` directory may be MIT.

    emit TokenUriUnpinned(tokenId, msg.sender);
  }

  // Checks if a token has a pinned URI.
  function hasPinnedTokenURI(uint256 tokenId) external view returns (bool isPinned) {
    require(msg.sender == ownerOf(tokenId), "Unauthorized");
    return _hasPinnedTokenURI[tokenId];
  }

  // Sets a specific metadata URI for a token at the given index.
  function setUri(uint256 tokenId, uint256 index, string calldata uri) external onlyOwner {
    if (_tokenURIs[tokenId].length > index) {
      _tokenURIs[tokenId][index] = uri;
    } else {
      _tokenURIs[tokenId].push(uri);
    }

    // Emit a MetadataUpdate event (see EIP-4906).
    emit MetadataUpdate(tokenId);
  }

  // Overrides supportsInterface to include IERC721MultiMetadata interface support.
  function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165, ERC721) returns (bool) {
    return (
      interfaceId == type(IERC721MultiMetadata).interfaceId ||
      super.supportsInterface(interfaceId)
    );
  }
}
```

## Security Considerations

This extension does not poses new security risks outside of those that might arise from existing token URI handling practices.

## Copyright

> This work is licensed under the MIT License.

Reviewer: EIPs MUST be CC0 licensed:

```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

## Motivation

The current [ERC-721](./eip-721.md) standard allows for a single metadata URI per token. However, there are use cases where multiple metadata URIs are desirable, such as when a token represents a collection of (cycling) assets with individual metadata, historic token metadata, collaborative and multi-artist tokens, evolving tokens. This extension enables such use cases by introducing the concept of multi-metadata support.

The primary reason for having a multi-metadata standard versus using the existing `tokenURI` method with some token-specific functions to change the active URI is that dapps and marketplaces don't have a mechanism to infer and display all the token URIs. Furthermore the non-pinned URIs might be equally important: imagine a colaborative token with multiple artworks per token. Multi-metadata tokens can be displayed in a carousel, gallery, list, etc.

## Specification

The `IERC721MultiMetadata` interface:

```solidity
interface IERC721MultiMetadata /* is IERC721 */ {
  event TokenUriPinned(uint256 indexed tokenId, uint256 indexed index, address indexed sender);
  event TokenUriUnpinned(uint256 indexed tokenId, address indexed sender);

>   function tokenURIs(uint256 tokenId) external view returns (uint256 index, string[] memory uris);

Author: @SamWilsn initially my idea was to propose only the addition of this method as it would allow dapps to retrieve all the uris and the current index (whether it is pinned or not).

However I wonder if it is a good idea to provide a standard interface to dapps/marketplaces to implement pinning, unpinning and related events. Should they decide to adopt this EIP I assume that it would make sense for them to have a standard way to offer pinning/unpinning via UI. By the way we'll try to get their opinion too.

Reviewer 2: @0xGh don't the functions on lines 34 and 35 accomplish this?

Author: Yes they do. Just sharing some context and thoughts based on what we discussed in private.

---
eip: 7160
title: ERC-721 Multi-Metadata Extension

> description: "Multiple metadata URIs per token, with the option to pin a primary URI."

Reviewer: ```suggestion
description: Multiple metadata URIs per token, with the option to pin a primary URI.
```

## Motivation

The current [ERC-721](./eip-721.md) standard allows for a single metadata URI per token with the `ERC721Metadata` implementation. However, there are use cases where multiple metadata URIs are desirable. Some example use cases are listed below:

- A token represents a collection of (cycling) assets with individual metadata
- An on-chain history of revisions to token metadata
- Appending metadata with different aspect ratios so that it can be displayed properly on all screens
- Dynamic and evolving metadata
- Collaborative and multi-artist tokens

This extension enables such use cases by introducing the concept of multi-metadata support.

The primary reason for having a multi-metadata standard in addition to the existing `ERC721Metadata` standard is that dapps and marketplaces don't have a mechanism to infer and display all the token URIs. Giving a standard way for marketplaces to offer collectors a way to pin/unpin one of the metadata choices also enables quick and easy adoption of this functionality.

> ## Specification

Reviewer 3: open
